### PR TITLE
[MODULAR] Dynamic threat/midround calculation

### PR DESCRIPTION
**Pull request by lessthnthree**
## About The Pull Request

The current way Dynamic is set up to allocate threat and roll midrounds makes no sense for how Skyrat operates. This adjusts the budgets and midround calculations.

When used in conjuction with a correct dynamic.json configuration the gamemode should work as expected.

## How This Contributes To The Skyrat Roleplay Experience

Dynamic operates like how you're trying to make it work. (No roundstart traitors, controlled midround rolls based on threat)

## Proof of Testing

<details>
<summary>Screenshots/Videos</summary>
  
image

image

[dynamic_simulations.json.log](https://github.com/Skyrat-SS13/Skyrat-tg/files/12033954/dynamic_simulations.json.log)

</details>

## Changelog

:cl: LT3
config: Dynamic gamemode rolls midrounds as expected
/:cl:

### CODE REVIEW

**Comment by chemistrymain2** [comment]
So what actually changes with this?

**Comment by PCPort** [comment]
No seriously what's actually being changed here

**Comment by lessthnthree** [comment]
This PR has no change to the dynamic.json config where the weights are set, but it does change how they're picked.

Heavy midrounds (blob) will roll as usual, just not at 2:50.

Midrounds that have been largely scarce (light pirates, nightmare, abductors etc.) will roll successfully more often.

```
/datum/game_mode/dynamic/generate_budgets()
	round_start_budget = 0
	initial_round_start_budget = 0
	mid_round_budget = threat_level
```
When the threat level for the round is determined, instead of portioning some off for roundstart traitors, 0 is given to roundstart and the entire threat_level budget is given to midrounds

```
/datum/game_mode/dynamic/get_heavy_midround_injection_chance(dry_run)
	var/next_midround_roll = next_midround_injection() - SSticker.round_start_time

	var/heavy_coefficient = CLAMP01((next_midround_roll - midround_light_upper_bound) / (midround_heavy_lower_bound - midround_light_upper_bound))

	return 100 * heavy_coefficient
```
When it comes time for a midround this formula is used to determine if it will be light or heavy. In short, the chance of the midround being 'heavy' increases with round time.

As an example from a current round on live at 1:25 and no midrounds have yet rolled:

image

If those numbers don't make sense, you'd be right! The heavy chance is going above 100%, which is meaningless.
It's not going to run a midround for another 1:22, which would put the round at 2:44 before it tries to roll any injection.
Right before the shuttle vote, hardly a midround. Also means the whole threat budget doesn't get spent.
In this example round admins would have to force a midround to make anything realistically happen.

This PR should fix that if used in conjunction with a new dynamic.json config, which I will provide to the server keyholders.

**Review by GoldenAlpharex — COMMENTED** [review]
Poke me if you'd like to have the current dynamics file, so it can be adjusted to take this into account.

**Comment by SomeRandomOwl** [comment]
so will this theoretically stop blobs and malf AI's from rolling in the last 20 minutes of the shift?

**Comment by lessthnthree** [comment]
Yes

File: modular_skyrat/master_files/code/game/gamemodes/dynamic.dm [hunk 1]
```diff
@@ -26,3 +22,31 @@
 	print_command_report(., "Central Command Status Summary", announce = FALSE)
 	priority_announce("Hello, crew of [station_name()]. Our intern has finished their shift-start divergency and goals evaluation, which has been sent to your communications console. Have a secure shift!", "Divergency Report", SSstation.announcer.get_rand_report_sound())
 
+/datum/game_mode/dynamic/
```
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/game/gamemodes/dynamic.dm]:
```suggestion
/datum/game_mode/dynamic
```

**Review by GoldenAlpharex — APPROVED** [review]
I'm willing to give this a shot, after tweaking Dynamic with this in mind.

File: modular_skyrat/master_files/code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm [hunk 2]
```diff
@@ -0,0 +1,4 @@
+/// Fired when there are no valid candidates. Try another roll after a delay.
+/datum/dynamic_ruleset/midround/from_ghosts/attempt_replacement()
+	mode.alternate_midround_injection()
+	return
```
Comment (GoldenAlpharex) [inline modular_skyrat/master_files/code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:4]: I like this change, thank you!

**Review by GoldenAlpharex — APPROVED** [review]

**Comment by GoldenAlpharex** [comment]
Should be all good to go!

**Pull request merged**